From a62b5d9e3516445bc10b9ff185123026b49ece85 Mon Sep 17 00:00:00 2001
From: Guillem Jover <guillem@debian.org>
Date: Fri, 26 Nov 2010 17:38:50 +0000
Subject: [PATCH 13/16] inetd: Change protocol semantics in inetd.conf

* src/inetd.c (setup): Update comment.
(getconfigent) [IPV6]: Change default family to IPv4 for "tcp" and
"udp". Change "tcp6" and "udp6" to support IPv4 mapped addresses.
Add "tcp6only" and "udp6only" protocols for IPv6 only addresses.
(fix_tcpmux) [IPV6]: Set se_proto to "tcp6".
[!IPV6]: Set se_proto to "tcp".
(cherry picked from commit bee2971bcd12559737aaa48132246fec74c384e7)

Backported-by: Simon McVittie <smcv@debian.org>
---
 inetd/inetd.c |   21 +++++++++++++++------
 1 files changed, 15 insertions(+), 6 deletions(-)

diff --git a/inetd/inetd.c b/inetd/inetd.c
index 5cd04cc..b9dde44 100644
--- a/inetd/inetd.c
+++ b/inetd/inetd.c
@@ -584,7 +584,11 @@ setup (struct servtab *sep)
 #ifdef IPV6
   if (sep->se_family == AF_INET6)
     {
-      int val = sep->se_v4mapped;
+      /* Reverse the value of SEP->se_v4mapped, since otherwise if
+	 using `tcp6' as a protocol type all connections will be mapped
+	 to IPv6, and with `tcp6only' they get mapped IPv4 mapped to
+	 IPv6.  */
+      int val = sep->se_v4mapped ? 0 : 1;
       if (setsockopt (sep->se_fd, IPPROTO_IPV6, IPV6_V6ONLY,
 		      (char *) &val, sizeof (val)) < 0)
 	syslog (LOG_ERR, "setsockopt (IPV6_V6ONLY): %m");
@@ -1137,10 +1141,9 @@ getconfigent (FILE *fconfig, const char *file, size_t *line)
       sep->se_proto = newstr (argv[INETD_PROTOCOL]);
 
 #ifdef IPV6
-      /* We default to IPv6, in setup() we'll fall back to IPv4 if
-         it doesn't work.  */
-      sep->se_family = AF_INET6;
-      sep->se_v4mapped = 1;
+      /* We default to IPv4.  */
+      sep->se_family = AF_INET;
+      sep->se_v4mapped = 0;
 
       if ((strncmp (sep->se_proto, "tcp", 3) == 0)
 	  || (strncmp (sep->se_proto, "udp", 3) == 0))
@@ -1148,7 +1151,11 @@ getconfigent (FILE *fconfig, const char *file, size_t *line)
 	  if (sep->se_proto[3] == '6')
 	    {
 	      sep->se_family = AF_INET6;
-	      sep->se_v4mapped = 0;
+	      /* Check for tcp6only and udp6only.  */
+	      if (strcmp (&sep->se_proto[3], "6only") == 0)
+	        sep->se_v4mapped = 0;
+	      else
+	        sep->se_v4mapped = 1;
 	    }
 	  else if (sep->se_proto[3] == '4')
 	    {
@@ -1361,9 +1368,11 @@ fix_tcpmux ()
       serv.se_fd = -1;			
       serv.se_type = NORM_TYPE;		
 #ifdef IPV6
+      serv.se_proto = newstr ("tcp6");
       serv.se_family = AF_INET6;
       serv.se_v4mapped = 1;
 #else
+      serv.se_proto = newstr ("tcp");
       serv.se_family = AF_INET;
 #endif
       if (debug)
-- 
1.7.2.3

